File: myshop/src/main/java/study/spring/myshop/order/command/domain/OrderNo.java
package study.spring.myshop.order.command.domain;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.Column;
import javax.persistence.Embeddable;
import java.io.Serializable;
import java.util.Objects;

@Embeddable
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
public class OrderNo implements Serializable {

    @Column(name = "order_number")
    private String number;

    public OrderNo(String number) {
        this.number = number;
    }

    @Override
    public int hashCode() {
        return Objects.hash(number);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;

        OrderNo other = (OrderNo)obj;
        return Objects.equals(number, other.number);
    }
}

File: 2nd/myshop/src/main/kotlin/study/ddd/myshop/order/domain/OrderNo.kt
package study.ddd.myshop.order.domain

import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
data class OrderNo (
    @Column(name = "order_number")
    val number: String
) : java.io.Serializable {

    companion object {
        fun of(number: String) : OrderNo = OrderNo(number)
    }
}